### ISSUE TITLE
Only analyze features included in model (predcontrib,predinteraction) 

### ISSUE THREAD

**Issue by palVJ** [shap/shap#868]
Having a large data set (in my case 350k x 450k), an ensemble method such as XGBoost may include after training only a small subset of features in the trees. At least when using R, using predict.xgb.Booster() with either predcontrib = TRUE or predinteractions = TRUE will return a matrix/array with all features from original training data (thereby consisting of a huge amount of zeros), but I only want to analyze those features included in the trees. 

I have looked at the source code (XGBoosterPredict_R -> XGBoosterPredict -> Predict -> PredictContribution/PredictInteractionContributions), but unfortunately extremely unexperienced with C++, and so not sure what should be done. Is there a quick fix? If someone would want to help me with this, I would deeply appreciate it. Simply setting model$feature_names = "FeaturesOfInterest" and making sure newdata/validation data only consists of "FeaturesOfInterest" in the R-function predict() does not work, where model is a xgb.Booster-object returned after training. It returns something very strange. I do think something needs to be done in source code. 
Also tried to use xgb.attr() or xgb.parameters<- with no success, but I think these are the ones that need to be used since it seems to change the variables within C++ environment.

In addition, as far as I have understood, the array when computing interactions may be very sparse. An idea would be to return an array something like a DelayedArray perhaps. See [https://github.com/Bioconductor/DelayedArray](url)

Related to #786

**Comment by slundberg** [comment]
An implementation of computing interactions for just a few rows is still not done as far I know. But in your case you could train an XGBoost model, then use the default global feature importances to select all the features that are non-zero (i.e. your model never used). Then retrain a new XGBoost model on just those features and then explain that model. This would get around SHAP not supporting sparse matrix return values right now.

**Comment by palVJ** [comment]
@slundberg, thank you very much. Yes, this is something I have been thinking about. Unfortunately, in my case this is not enough. As an example with 110 000 features, the final model consists of 6000 features in total. In this particular analysis, it is important to have a sufficiently large validation set because of the heterogeneity in the data. Say I would want to have a validation set of 1000 samples (I would like much more). If I would want to analyze the interactions, the model (in R) would return an array of size (6000x6000x1000). This will for instance require 134 GB in R.

In addition, as far as I have understood though I may be completely wrong, due to the structure and dependencies of the features in the trees, you will not have for each individual shap-interaction values equal to (6000 choose 2 = 18 mill.), but possibly only a small subset of this again. Am I right?

However, one could do what you said in a repetitive manner, say first with 110 000 features, then with 6000 features etc. But then you rather do it for the sake of memory reduction more than for the sake of analysis, and you actually must rely on that you end up with a small enough subset.
Do you think implementation for only picking those features you want to analyze will be time-consuming? I would love to help if ever needed.

**Comment by slundberg** [comment]
Yes the matrix is still very sparse then (unless you have a million trees), so sparse support would still be better. But as a stop gap you don't need to explain all 1k samples at the same time. You can explain one sample at a time (which would be slow).

You could also raise the threshold above zero to get less than 6k features. To make it support sparse data would require a fair amount of C++ work which I don't have the bandwidth for at the moment. Help would be appreciated there, but it is important to keep in mind it is not a small task.

**Comment by palVJ** [comment]
aha, thank you for clarifying for me. My first PhD-paper will most likely rely on computations of shap-values, so I will be more than happy to help.
